walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,169 @@
+        } else {
+          break;
+        }
+        continue;
+      }
+      // There is another {@code TimeRange} that ends later than the current ending
+      // {@code TimeRange} + before the immediately following, starting {@code TimeRange}.
+      // (Since there exists a {@code TimeRange} starting later than the current ending
+      //  {@code TimeRange}, there must be other {@code TimeRange} ending even later. Therefore,
+      //  the {@code endOrderedPointer} is guaranteed to not go out of range.)
+      TimeRange nextEndingTimeRange = endOrderedTimeRangesIterator.next();
+      if (nextEndingTimeRange.end() <= startingTimeRange.start()) {
+        endingTimeRange = nextEndingTimeRange;
+        continue;
+      } else {
+        endOrderedTimeRangesIterator.previous();
+      }
+
+      // At this point, we have:
+      //    |----ending {@code TimeRange}----| available time |----starting {@code TimeRange}----|
+      int availableTimeRangeStart = endingTimeRange.end();
+      int availableTimeRangeEnd = startingTimeRange.start();
+      checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                            availableTimeRangeStart, availableTimeRangeEnd,
+                                            availableTimeRanges);
+
+      // Find the potential, next block of available time sandwiched between an ending and a
+      // starting {@code TimeRange}.
+      // (We can safely increment both pointers without missing any available time, because
+      //  we had the guarantee that no other existing {@code TimeRange} can be found between
+      //  the current ending and starting {@code TimeRange}.)
+      if (startOrderedTimeRangesIterator.hasNext() && endOrderedTimeRangesIterator.hasNext()) {
+        startingTimeRange = startOrderedTimeRangesIterator.next();
+        endingTimeRange = endOrderedTimeRangesIterator.next();
+      } else {
+        break;
+      }
+    }
+
+    // Potentially add the {@code TimeRange} after the last existing {@code TimeRange}.
+    TimeRange lastEndingTimeRange = endOrderedTimeRanges.get(endOrderedTimeRanges.size() - 1);
+    checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                          lastEndingTimeRange.end(), TimeRange.END_OF_DAY + 1,
+                                          availableTimeRanges);
+
+    return availableTimeRanges;
+  }
+
+  /**
+  * Finds the existing {@code TimeRange} that correspond to {@code Event} having
+  * overlapping attendees with a set of attendees that we are concerned with.
+  */
+  private Collection<TimeRange> getRelevantTimeRangesFromEvents(Collection<Event> events,
+      Set<String> attendees) {
+    Collection<TimeRange> relevantTimeRanges = new HashSet<>();
+    for (Event event : events) {
+      Set<String> overlappingAttendees = new HashSet<>(event.getAttendees());
+      overlappingAttendees.retainAll(attendees);
+      if (!overlappingAttendees.isEmpty()) {
```

Comment:
You could consider using https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#disjoint(java.util.Collection,%20java.util.Collection) so that you don't have to make additional copies of the read-only attendees.